crates/handlers/src/views/recovery/start.rs

```diff
@@ -0,0 +1,137 @@
+// Copyright 2024 The Matrix.org Foundation C.I.C.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+use std::str::FromStr;
+
+use axum::{
+    extract::State,
+    response::{Html, IntoResponse, Response},
+    Form, TypedHeader,
+};
+use lettre::Address;
+use mas_axum_utils::{
+    cookies::CookieJar,
+    csrf::{CsrfExt, ProtectedForm},
+    FancyError, SessionInfoExt,
+};
+use mas_data_model::UserAgent;
+use mas_router::UrlBuilder;
+use mas_storage::{BoxClock, BoxRepository, BoxRng};
+use mas_templates::{
+    FieldError, FormState, RecoveryStartContext, RecoveryStartFormField, TemplateContext, Templates,
+};
+use serde::{Deserialize, Serialize};
+
+use crate::{BoundActivityTracker, PreferredLanguage};
+
+#[derive(Deserialize, Serialize)]
+pub(crate) struct StartRecoveryForm {
+    email: String,
+}
+
+pub(crate) async fn get(
+    mut rng: BoxRng,
+    clock: BoxClock,
+    mut repo: BoxRepository,
+    State(templates): State<Templates>,
+    State(url_builder): State<UrlBuilder>,
+    PreferredLanguage(locale): PreferredLanguage,
+    cookie_jar: CookieJar,
+) -> Result<Response, FancyError> {
+    let (session_info, cookie_jar) = cookie_jar.session_info();
+    let (csrf_token, cookie_jar) = cookie_jar.csrf_token(&clock, &mut rng);
+
+    let maybe_session = session_info.load_session(&mut repo).await?;
+    if maybe_session.is_some() {
+        // TODO: redirect to continue whatever action was going on
+        return Ok((cookie_jar, url_builder.redirect(&mas_router::Index)).into_response());
+    }
+
+    let context = RecoveryStartContext::new()
+        .with_csrf(csrf_token.form_value())
+        .with_language(locale);
+
+    repo.save().await?;
```

Comment:
why save?

Reply:
I did read somewhere that `commit`ing a transaction, even if it was read only, was better than rolling it back? (which is effectively what `Drop`ping the repository will do). Although you're right that this isn't consistant with anything. Should I add `save`s everywhere or drop this one?

Reply:
that was sort of part of my point yes, we have some early returns where we don't save.

There is a thing in Postgres, at least under the SERIALIZABLE isolation level, where when you read something, you can't trust it until you finish your transaction, commit and it commits successfully. I thought I wrote a blog post explaining the mechanics of this but it seems I don't have it. Maybe it's unpublished on a different computer :(.

The high-level summary from Postgres' manual is:

> When relying on Serializable transactions to prevent anomalies, it is important that any data read from a permanent user table not be considered valid until the transaction which read it has successfully committed. This is true even for read-only transactions, except that data read within a deferrable read-only transaction is known to be valid as soon as it is read, because such a transaction waits until it can acquire a snapshot guaranteed to be free from such problems before starting to read any data. In all other cases applications must not depend on results read during a transaction that later aborted; instead, they should retry the transaction until it succeeds.
> 
> — https://www.postgresql.org/docs/current/transaction-iso.html

Annoyingly I don't think MAS is well-structured for this. (I'm not sure how you can do this ergonomically without having to structure your DB code like Synapse does: where you pass a transaction function and Synapse retries it as needed.)
But I'm not sure you use the serialisable transaction isolation level anyway and I can't remember if repeatable read has the same caveat for read-only transactions.


I am not sure which side of the fence to sit on here, but I don't think it's easy to remember to save before every early exit. :/

Reply:
We're indeed not using the SERIALIZABLE isolation level. Most of the time if we do an early exit, it's because we're in an error state anyway, so it probably doesn't matter? Not sure what to do here

Reply:
yeah. Maybe we just leave it as is. It's not super satisfying but not something to get stuck on too much maybe